Layout the snek gud

oheyjesse opened the pull request:
<img width="690" alt="Screen Shot 2019-08-07 at 17 04 43" src="https://user-images.githubusercontent.com/26806428/62601810-996d6780-b935-11e9-9bc8-69368b1db929.png">

# What's Snekkenin'?

This is a rough PR to put in some colour, some spacing, some layout goodness. I was going to mess with it more, but webpack and Css bundling and all that fun stuff got in the way. :( 

## Notable changes:

### Structure / CSS / JS builds etc
- Added `main.js` to gitignore. It gets built by webpack all the time so... Let's not bother committing it.

- Deleted `farms/static/css/snek.css` - We had this file in `frontend/static/css/snek.css` and webpack wasn't bundling both. The CSS management in here needs a bunch of work (webpack config not properly set up) but this meant I could add / change styled in `frontend/` rather than the `farms/` model, which seemed to make more sense for now (in a global styles sense).

- Css should probably be built and output by webpack, anway, which would mean we could source it from multiple files if set up correctly, or even go with some kind of `styled-components` css-in-js system, but I didnt go that far :P I just merged what we had into one file in `frontend/static`. Webpack doesn't build it.

### Component Shuffling / New stuff
- Moved `Snek.js` and all related `Snektions` to their own `Snek/` component folder. Cleans up the components directory a bit and makes for easier navigation

- Pretty heavily reworked the `<divs>` inside the main `Farm` render - Put them all inside a main Flexbox container and labelled them accordingly. Three sections: Buttons, LongGrass (Visualisations), and SnekBoard (Dashboard - shows snek count numbers etc)

- Converted the `micropython_array` func to a <Micropythons/> component, and the same with `curlyboi` -> `<Curlyboi/>` to keep in line with React conventions

Commit:
:art: Move the Snek into it's own Snekky Folder

Commit:
:art: Do flexbox layouts and colors - move css into frontend/static for simplicity

Commit:
:art: Post PR Fixes/Tidyup

mermop reviewed (approved):
~Changes generally look fine but there's a bit of duplication with #18 - happy to merge but it might be a good idea to tidy those up!~ Review left on the wrong PR >_<

mermop commented on frontend/src/components/Farm.js:
```diff
@@ -1,61 +1,81 @@
+        </section>
+
+        <section className="farm farm--longGrass">
+          <Micropythons/>
```
👍 Good refactor here

mermop commented on frontend/src/components/Farm.js:
```diff
@@ -1,61 +1,81 @@
+          <Curlyboi/>
+        </section>
+
+        <section className="farm farm--snekBoard">
```
What's the thought behind this class name?

oheyjesse replied:
Honestly, I don't know CSS Naming and BEM very well, and was just giving it my best shot 😂 I'll take a look at the BEM Specifications and give this another go! I'm mostly fumbling around in the dark when it comes to CSS - most my experience is with CSS-in-JS which is a different beast, and doesn't use as much naming.

I think my overall naming structure was to try to have things labelled as "farm" since they're sub-elements of the "farm" container, but perhaps I went overboard with the `--`s, and also, now that I think of it, giving them all the class `farm` kind of defeats the point, when I can instead target "all farm elements" using CSS Selectors etc. I'll rethink this

mermop commented on frontend/static/css/snek.css:
```diff
@@ -1,11 +1,61 @@
+}
+
+/* Root Level Container */
+.wrapper--container {
```
Usually the two hyphens `--` denote a BEM block with a modifier `block--modifier` but it doesn't look like that's the case here - can you explain your thinking?

oheyjesse replied:
This could just be `wrapper` - I'm not sure why I put in `--container`! I don't know if I fully thought this through.

mermop commented on frontend/static/css/snek.css:
```diff
@@ -1,11 +1,61 @@
+  background-color: var(--purple);
+}
+
+.farm > section {
```
A BEM approach here would be to give the sections a class like `farm__section` and target them that way - that way you don't have to be tied to the `section` element. Makes things more flexible

oheyjesse replied:
Good tip! I will do that. Thanks for the advice.

mermop commented on frontend/static/css/snek.css:
```diff
@@ -1,11 +1,61 @@
+  background-color: var(--color-bg-alt);
+}
+
+.farm--buttons {
```
If you're going with BEM this should be `farm__buttons`

oheyjesse replied:
I'll take a read through some guidelines re: all the naming. I'm thinking https://cssguidelin.es/#bem-like-naming looks like a good place to start?

mermop replied:
That one looks good - also, http://getbem.com/naming/ is the definitive source

mermop commented on frontend/static/css/snek.css:
```diff
@@ -1,11 +1,61 @@
 :root {
+  /* Defaults */
+  --color-bg-main: #ffed4d;
+  --color-bg-alt: #bee7c6;
```
Can I be real here, I think these colours are a bit Too Much

oheyjesse replied:
HAhahaha yes, that was definitely an "Eh, I'll see what they think" approach. More than happy to tone it down, or even leave it to others. I can remove the borders/colours and we can hash out a particular design first, if you like :)

mermop replied:
I reckon the general background definitely does not need to be yellow - the purple sections are probably fine... we can play around with it more later :)

oheyjesse replied:
Eh, I'll go back to no background colours for now, for the `Full Colour Snek Effeckt`

mermop commented on frontend/src/components/Farm.js:
```diff
@@ -1,61 +1,81 @@
+
+        <section className="farm farm--snekBoard">
+          <p>Micropythons: {micropythons}</p>
+          <p>Curlyboi Length: {curlyboi_snektions}</p>
```
I liked the flavour text of your curlyboi cultivation effort

oheyjesse replied:
Thought about that concern too, and didn't want to lose it, but from a layout perspective, where do you think it's best placed? I thought about a "message" section, that could pop up fun / helpful messages as you reach certain points, but I thought that'd be more of a feature to implement. I was mostly considering mobile view, here- thinking about page width etc, and how it'd look when more properties are added to the dashboard/item count

mermop replied:
How about underneath 'welcome to <farm name> farm' for the moment?

mermop reviewed (changes requested):
A couple of comments:

- There is some duplication with #19 
- Class names can be more consistent
- I am not convinced about the bright coloured background!! I think it detracts from the glory and delight of the bois

Refactoring of JS looks great though

oheyjesse commented:
Thanks a lot @mermop ! I didn't expect such a thorough review, and I really appreciate it. 

I addressed a few things in the individual comments and am happy to rework stuff. Absolutely solid points on the CSS naming and even the colours / layout structure itself. I took some liberties, and would love to work to get there.

I'm not sure the best approach to layout / design now that we're all in separate spaces. Happy to talk out more of a plan rather than coding ahead too far :)

Commit:
:recycle: Roll back .gitignore and webpack.config.js

Commit:
:art: Fix up naming in layout components, tidy up some erroneous rebase/merge leftovers

mermop reviewed (approved):
Looks good to me 👍🐍

libbyberrie commented:
This is fantastic Drew! It looks great to me! ✨ 🐍

mermop commented:
Sorry for the conflicts from #23 😅

oheyjesse commented:
Haha, all good- I will get back to it later :P

libbyberrie committed:
Fix most merge conflicts

libbyberrie committed:
define up in initialstate and Snek object

libbyberrie committed:
Add main,js to gitignore

mermop commented on .gitignore:
```diff
@@ -112,3 +112,6 @@ venv.bak/
 .vscode
+
+# Webkit's compiling
+frontend/static/frontend/main.js
```
⚠️ This needs to be checked in for the deployment to Heroku to work

oheyjesse replied:
Yeah, we had to change our stance on the `main.js` recently due to deployment. For now, it stays.

libbyberrie replied:
Ahh cool. I'll remove that from gitignore then.

mermop commented on frontend/src/components/App.js:
```diff
@@ -6,7 +6,10 @@ import Farm from "./Farm";
     endpoint="api/farm"
-    render={data => <Farm farm={data[0]}/>}
+    render={data => (
+      <div className="wrapper--container">
```
This class name isn't used

oheyjesse replied:
Ah, I believe it's just `wrapper` now. My bad!

libbyberrie committed:
🔥 remove skekvenv from repo

libbyberrie committed:
Add snekvenv to gitignore, remove main.js from gitignore for heroku.

libbyberrie committed:
updated main.js

oheyjesse commented on .gitignore:
```diff
@@ -87,6 +87,7 @@ celerybeat-schedule
 .venv
 env/
 venv/
+snekvenv/
```
I love this haha

oheyjesse commented on frontend/src/components/Farm.js:
```diff
@@ -42,24 +47,49 @@ const Farm = ({ farm }) => {
     micropython_array.push(<Snek size="micropython" length={5} up={i%2==0 ? state.up : !state.up}/>)
```
This whole section from the console log including "micropython_array" is duplicated - I think it's come back in the merge :/

libbyberrie replied:
Sorry, my bad! Will remove after work.

oheyjesse commented on frontend/src/components/Farm.js:
```diff
@@ -42,24 +47,49 @@ const Farm = ({ farm }) => {
+    )
+  }
+
   if(state.curlyboi_snektions > 0) {
```
This `if` is old code too that's duplicated inside "curlyboi"

libbyberrie replied:
Thanks for spotting that -- i'm surprised i didn't see it before D:

oheyjesse reviewed (commented):
I had a quick fire-up of this one this morning, and sadly, there's 

- ⚠️ It looks like the animations are broken in this build :( Nuuuuuuuu! It... might be CSS class stuff? Not sure...

- Some duplicated code, have commented in the review (must've come through in the rebase/merge)

mermop commented:
The CSS all looks fine - it's got the `transition` correctly. I suspect that the sneks are not animating because they don't consider themselves to be the same element - so it's something to do with the refactor of the JS.

libbyberrie committed:
🔥remove duplicate declarations of curlyboi and micropythons

libbyberrie committed:
Change units to snektions. Update main.js

libbyberrie commented on frontend/src/components/Snek/SnekN.js:
```diff
@@ -1,20 +1,27 @@
+      stroke-linecap="round"
+      stroke="#5b57a6"
+      fill="transparent"
+      class="snake-body"
```
@mermop I noticed you changed most of the classNames from the SVG path into 'class', except for SnekEnd.js. Did you want me to change that to match also?

Commit:
:wrench: Fix The Animations! Add A Bigger Button! Do It All In One Big Stupid Commit!

Sorry frens

Commit:
:art: Lil refactor on some of the logic, some more comments to clarify certain points

Commit:
:beers: I did a late night oops. Fixed the late night oops

oheyjesse commented:
![snekanim](https://user-images.githubusercontent.com/26806428/63344359-9b9edf80-c393-11e9-95cc-a8eb2d96baa3.gif)

THE SNEK LIVES! And `D A N C E S`
- Also made the button bigger for `MOAR MASHING ON MOBILE`

So, as it turns out, React does some pretty intense rendering things when it detects changes to a component. If, as is in our app, something is controlled via the component's internal `state`, react will specifically attempt to `re-render` that thing inside of the component. 

However, when I refactored the logic out to their own _separate_ functional components with `const Micropythons = () => { ... }` - I was inadvertently telling React to _completely re-render these components_ whenever their internal parts changed. This would wipe out and re-render all the sneks at once, which would break all their lovely transitions!

There were two ways around this: 
One: to go in and give the `<Snek>` or even `<Snektions>` components their own pieces of state, and then have them immediately update that state based on what was passed in. I attempted this, and it didn't really work. It's not a good workaround- it was causing waaay too many events to fire (one for each snektion) and isn't at all scalable. It was a pretty good example of "what not to do with state" - the results were _not pretty_ 😂 

Two: The other way to fix it was to go back to how we had it before; that is- have the main "farm" component manage the state, and pass it down to the children. I had to revert it back to not wrapping the `micropythons` and `curlyboi` in their own components, and rather have our main `Farm` component do the logic, based on it's own state. This lets react do it's rendering magic and merely modify the classes rather than re-rendering everything.

This does limit us a little in that we can't "extract out" the logic sections for rendering the micropythons or curlybois, but for now, it works. I thought about making a `Micropythons` component and `Curlyboi` component, and give each of __those__ a piece of state to manage themselves, but I ended up just going 🤷‍♂ and leaving it as it was.

I also did some refactors - the `updateSneks` function was a little long and verbose- I had a go at drying it up a little by calculating each updated state piece once, then updating the state once.

I also refactored the micropython render `for` loop into an array the size of the `micropythons` number, and mapped over it. This was mostly personal stylistic preference- I wasn't huge on the `for`, and prefer to keep to a `.map` syntax - I'm a bigger fan of the "pass a function to `.map`" style of javascript over using for loops - But I'm happy to reneg on this if it's not y'all's preference)

libbyberrie reviewed (approved):
💪 
Are you happy for me to merge this? It's looking fantastic!
